File: backend/app/routers/cabs.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
from .. import models, schemas
from ..database import get_db

router = APIRouter(prefix="/cabs", tags=["cabs"])

@router.get("/routes", response_model=List[schemas.CabRoute])
def get_routes(db: Session = Depends(get_db)):
    """Get all available cab routes"""
    routes = db.query(models.CabRoute).filter(models.CabRoute.is_active == True).all()
    return routes

@router.get("/vehicles", response_model=List[schemas.Vehicle])
def get_vehicles(db: Session = Depends(get_db)):
    """Get all available vehicle types"""
    vehicles = db.query(models.Vehicle).filter(models.Vehicle.is_active == True).all()
    return vehicles

@router.post("/search", response_model=List[schemas.CabSearchResult])
def search_cabs(
    search: schemas.CabSearchRequest,
    db: Session = Depends(get_db)
):
    """
    Search for available cabs between two locations.
    Returns all vehicle options with pricing for the route.
    """
    # Find the route
    route = db.query(models.CabRoute).filter(
        models.CabRoute.from_location.ilike(f"%{search.from_location}%"),
        models.CabRoute.to_location.ilike(f"%{search.to_location}%"),
        models.CabRoute.is_active == True
    ).first()
    
    if not route:
        raise HTTPException(status_code=404, detail="No route found between these locations")
    
    # Get all pricing for this route
    pricing_list = db.query(models.CabPricing).filter(
        models.CabPricing.route_id == route.id
    ).all()
    
    # Build results
    results = []
    for pricing in pricing_list:
        vehicle = db.query(models.Vehicle).filter(
            models.Vehicle.id == pricing.vehicle_id,
            models.Vehicle.is_active == True
        ).first()
        
        if vehicle:
            results.append({
                "route": route,
                "vehicle": vehicle,
                "price": pricing.price
            })
    
    return results

@router.post("/routes", response_model=schemas.CabRoute)
def create_route(route: schemas.CabRouteCreate, db: Session = Depends(get_db)):
    """Create a new cab route (Admin only - auth to be added later)"""
    db_route = models.CabRoute(**route.dict())
    db.add(db_route)
    db.commit()
    db.refresh(db_route)
    return db_route

@router.post("/vehicles", response_model=schemas.Vehicle)
def create_vehicle(vehicle: schemas.VehicleCreate, db: Session = Depends(get_db)):
    """Create a new vehicle type (Admin only - auth to be added later)"""
    db_vehicle = models.Vehicle(**vehicle.dict())
    db.add(db_vehicle)
    db.commit()
    db.refresh(db_vehicle)
    return db_vehicle

File: backend-node/src/routes/cabs.js
const express = require("express");
const router = express.Router();
const prisma = require("../db");

// GET /cabs/routes
router.get("/routes", async (req, res, next) => {
  try {
    const routes = await prisma.cabRoute.findMany({
      where: { isActive: true },
      include: { pricing: { include: { vehicle: true } } },
    });
    res.json(routes);
  } catch (err) {
    next(err);
  }
});

// GET /cabs/vehicles
router.get("/vehicles", async (req, res, next) => {
  try {
    const vehicles = await prisma.vehicle.findMany({
      where: { isActive: true },
    });
    res.json(vehicles);
  } catch (err) {
    next(err);
  }
});

// POST /cabs/search  { from_location, to_location }
router.post("/search", async (req, res, next) => {
  try {
    const { from_location, to_location } = req.body;

    const route = await prisma.cabRoute.findFirst({
      where: {
        fromLocation: { contains: from_location, mode: "insensitive" },
        toLocation: { contains: to_location, mode: "insensitive" },
        isActive: true,
      },
    });

    if (!route) {
      return res.status(404).json({ detail: "No route found between these locations" });
    }

    const pricingList = await prisma.cabPricing.findMany({
      where: { routeId: route.id },
      include: { vehicle: true },
    });

    const results = pricingList
      .filter((p) => p.vehicle.isActive)
      .map((p) => ({
        route,
        vehicle: p.vehicle,
        price: p.price,
      }));

    res.json(results);
  } catch (err) {
    next(err);
  }
});

// POST /cabs/routes
router.post("/routes", async (req, res, next) => {
  try {
    const { fromLocation, toLocation, distanceKm, durationHours, isActive = true } = req.body;

    const route = await prisma.cabRoute.create({
      data: { fromLocation, toLocation, distanceKm, durationHours, isActive },
    });
    res.status(201).json(route);
  } catch (err) {
    next(err);
  }
});

// POST /cabs/vehicles
router.post("/vehicles", async (req, res, next) => {
  try {
    const { name, capacity, luggageCapacity, features, isActive = true } = req.body;

    const vehicle = await prisma.vehicle.create({
      data: { name, capacity, luggageCapacity, features, isActive },
    });
    res.status(201).json(vehicle);
  } catch (err) {
    next(err);
  }
});

module.exports = router;
